execution/gethexec/executionengine.go

```diff
@@ -275,6 +275,59 @@ func (s *ExecutionEngine) SequenceTransactions(header *arbostypes.L1IncomingMess
 	})
 }
 
+func (s *ExecutionEngine) SequenceMessage(msg arbostypes.L1IncomingMessage, hooks *arbos.SequencingHooks) (*types.Block, error) {
+	return s.sequencerWrapper(func() (*types.Block, error) {
+		lastBlockHeader, err := s.getCurrentHeader()
+		if err != nil {
+			return nil, err
+		}
+
+		statedb, err := s.bc.StateAt(lastBlockHeader.Root)
+		if err != nil {
+			return nil, err
+		}
+
+		delayedMessagesRead := lastBlockHeader.Nonce.Uint64()
+
+		startTime := time.Now()
+		block, receipts, err := arbos.ProduceBlock(
+			&msg,
+			delayedMessagesRead,
+			lastBlockHeader,
+			statedb,
+			s.bc,
+			s.bc.Config(),
+			func(batchNum uint64) ([]byte, error) { return nil, errors.New("invalid tx") },
+		)
+		blockCalcTime := time.Since(startTime)
+
+		pos, err := s.BlockNumberToMessageIndex(lastBlockHeader.Number.Uint64() + 1)
+		if err != nil {
+			return nil, err
+		}
+
+		msgWithMeta := arbostypes.MessageWithMetadata{
+			Message:             &msg,
+			DelayedMessagesRead: delayedMessagesRead,
+		}
+
+		err = s.streamer.WriteMessageFromSequencer(pos, msgWithMeta)
```

Comment:
IIUC, here we are sending `msg` to L1. `msg` contains all the transactions including invalid ones, even if some invalid transactions were filtered out by the hooks in `ProduceBlock(Advanced)`?

Reply:
That's why the `SequenceMessage` (I think we should call it `SequenceTransactionsEspresso` instead) function was added. It's doing what the sequencer was doing but not applying the hooks after producing a block.